test/e2e/aks_webhooks.go

```diff
@@ -0,0 +1,391 @@
+	amcp, err = getAzureManagedControlPlane(ctx, mgmtClient, input.Cluster.Namespace, input.Cluster.Name)
+	Expect(err).NotTo(HaveOccurred())
+	amcp.Spec.ResourceGroupName = amcp.Spec.ResourceGroupName + "foo"
+	err = mgmtClient.Update(ctx, amcp)
+	Expect(err).To(HaveOccurred())
+	Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+	By("Validating that we can't update AzureManagedControlPlane.Spec.NodeResourceGroupName")
+	amcp, err = getAzureManagedControlPlane(ctx, mgmtClient, input.Cluster.Namespace, input.Cluster.Name)
+	Expect(err).NotTo(HaveOccurred())
+	amcp.Spec.NodeResourceGroupName = amcp.Spec.NodeResourceGroupName + "foo"
+	err = mgmtClient.Update(ctx, amcp)
+	Expect(err).To(HaveOccurred())
+	Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+	By("Validating that we can't update AzureManagedControlPlane.Spec.Location")
+	amcp, err = getAzureManagedControlPlane(ctx, mgmtClient, input.Cluster.Namespace, input.Cluster.Name)
+	Expect(err).NotTo(HaveOccurred())
+	amcp.Spec.Location = "notavalidregion"
+	err = mgmtClient.Update(ctx, amcp)
+	Expect(err).To(HaveOccurred())
+	Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+	By("Validating that we can't update AzureManagedControlPlane.Spec.SSHPublicKey")
+	amcp, err = getAzureManagedControlPlane(ctx, mgmtClient, input.Cluster.Namespace, input.Cluster.Name)
+	Expect(err).NotTo(HaveOccurred())
+	amcp.Spec.SSHPublicKey = amcp.Spec.SSHPublicKey + "foo"
+	err = mgmtClient.Update(ctx, amcp)
+	Expect(err).To(HaveOccurred())
+	Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+	By("Validating that we can't update AzureManagedControlPlane.Spec.DNSServiceIP")
+	amcp, err = getAzureManagedControlPlane(ctx, mgmtClient, input.Cluster.Namespace, input.Cluster.Name)
+	Expect(err).NotTo(HaveOccurred())
+	amcp.Spec.DNSServiceIP = to.StringPtr("256.256.256.256")
+	err = mgmtClient.Update(ctx, amcp)
+	Expect(err).To(HaveOccurred())
+	Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+	By("Validating that we can't update AzureManagedControlPlane.Spec.NetworkPlugin")
+	amcp, err = getAzureManagedControlPlane(ctx, mgmtClient, input.Cluster.Namespace, input.Cluster.Name)
+	Expect(err).NotTo(HaveOccurred())
+	if to.String(amcp.Spec.NetworkPlugin) == "azure" {
+		amcp.Spec.NetworkPlugin = to.StringPtr("kubenet")
+	} else {
+		amcp.Spec.NetworkPlugin = to.StringPtr("azure")
+	}
+	err = mgmtClient.Update(ctx, amcp)
+	Expect(err).To(HaveOccurred())
+	Expect(apierrors.IsInvalid(err)).To(BeTrue())
+
+	By("Validating that we can't update AzureManagedControlPlane.Spec.NetworkPolicy")
+	amcp, err = getAzureManagedControlPlane(ctx, mgmtClient, input.Cluster.Namespace, input.Cluster.Name)
+	Expect(err).NotTo(HaveOccurred())
+	if to.String(amcp.Spec.NetworkPolicy) == "azure" {
+		amcp.Spec.NetworkPolicy = to.StringPtr("calico")
+	} else {
+		amcp.Spec.NetworkPolicy = to.StringPtr("azure")
+	}
+	err = mgmtClient.Update(ctx, amcp)
+	Expect(err).To(HaveOccurred())
+	Expect(apierrors.IsInvalid(err)).To(BeTrue())
```

Comment:
I wonder if a lot of the boilerplate could be removed with an extra layer of `BeforeEach`/`AfterEach`, if Ginkgo allows that here. I imagine that might be able to turn most of this into something like:

```go
	var amcp *infrav1exp.AzureManagedControlPlane

	BeforeEach(func() {
		amcp, err = getAzureManagedControlPlane(ctx, mgmtClient, input.Cluster.Namespace, input.Cluster.Name)
		Expect(err).NotTo(HaveOccurred())
	})

	AfterEach(func() {
		err = mgmtClient.Update(ctx, amcp)
		Expect(err).To(HaveOccurred())
		Expect(apierrors.IsInvalid(err)).To(BeTrue())
	})

	It("fails when updating Spec.NetworkPlugin", func() {
		if to.String(amcp.Spec.NetworkPlugin) == "azure" {
			amcp.Spec.NetworkPlugin = to.StringPtr("kubenet")
		} else {
			amcp.Spec.NetworkPlugin = to.StringPtr("azure")
		}
	})
```

Certainly not necessary to address now, just an idea.